src/main/java/se/liu/ida/hefquin/engine/queryproc/impl/optimizer/evolutionaryAlgorithm/CostModel.java

```diff
@@ -0,0 +1,304 @@
+import se.liu.ida.hefquin.engine.queryplan.logical.impl.LogicalOpTPAdd;
+import se.liu.ida.hefquin.engine.queryplan.physical.PhysicalOperatorForLogicalOperator;
+import se.liu.ida.hefquin.engine.queryplan.physical.impl.*;
+import se.liu.ida.hefquin.engine.queryplan.utils.ExpectedVariablesUtils;
+import se.liu.ida.hefquin.engine.queryproc.QueryOptimizationException;
+import se.liu.ida.hefquin.engine.queryproc.QueryProcContext;
+import se.liu.ida.hefquin.engine.queryproc.impl.optimizer.utils.CardinalityEstimation;
+
+import java.util.*;
+
+public class CostModel extends CardinalityEstimation{
+    protected final QueryProcContext ctxt;
+    protected final RootOperatorCostCache rootOperatorCostCache = new RootOperatorCostCache();
+
+    public CostModel(QueryProcContext ctxt) {
+        super(ctxt);
+        this.ctxt = ctxt;
+    }
+
+    public Double overallCost( final PhysicalPlan pp ) throws QueryOptimizationException {
+        List<Integer> measureMetrics = costMetrics( pp, new ArrayList<Integer>() );
+        ArrayList<Double> weight= new ArrayList<Double>(Arrays.asList(0.2, 0.2, 0.2, 0.2, 0.2));
+
+        double overallCost = 0;
+        for (  int i = 0; i < measureMetrics.size(); i++ ){
+            overallCost = overallCost + measureMetrics.get(i) * weight.get(i);
+        }
+
+        return overallCost;
+    }
+
+    protected List<Integer> costMetrics( final PhysicalPlan pp, final List<Integer> overallMetrics ) throws QueryOptimizationException {
+        final List<PhysicalPlan> children = findChildren(pp);
+
+        for ( int i = 0; i < children.size(); i++ ){
+            final List<Integer> li = metricsOfRootOperator( children.get(i) );
+            for ( int j = 0; j < li.size(); j++ ){
+                overallMetrics.set( j, overallMetrics.get(j) + li.get(j) ) ;
+            }
+            costMetrics( children.get(i), overallMetrics);
+        }
+
+        return overallMetrics;
+    }
+
+    protected List<PhysicalPlan> findChildren(final PhysicalPlan pp ) throws QueryOptimizationException {
+        final List<PhysicalPlan> children = new ArrayList<PhysicalPlan>();
+        final int numChildren = pp.numberOfSubPlans();
+
+        for ( int i = 0; i < numChildren; ++i ){
+            children.add( pp.getSubPlan(i) );
+        }
+        return children;
+    }
+
+    protected List<Integer> metricsOfRootOperator(final PhysicalPlan pp ) throws QueryOptimizationException {
+        final List<Integer> metrics = new ArrayList<Integer>();
+
+        metrics.add( numberOfRequests( pp.getRootOperator(), pp ) );
+        metrics.add( shippedRDFTermsForRequests( pp.getRootOperator(), pp ) );
+        metrics.add( shippedRDFVarsForRequests( pp.getRootOperator(), pp ) );
+        metrics.add( shippedRDFTermsForResponses( pp.getRootOperator(), pp ) );
+        metrics.add( shippedRDFVarsForResponses( pp.getRootOperator(), pp ) );
+
+        rootOperatorCostCache.add( pp, metrics );
```

Comment:
If you add something to this cache here, shouldn't you also try to find it in the cache at the begin of this method?